src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,140 @@
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using OmniSharp.Extensions.JsonRpc;
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal abstract class AbstractRazorDelegatingEndpoint<TRequest, TResponse, TDelegatedParams> : IJsonRpcRequestHandler<TRequest, TResponse?>
+        where TRequest : TextDocumentPositionParams, IRequest<TResponse>
+        where TResponse : class
+        where TDelegatedParams : class
+    {
+        private readonly DocumentContextFactory _documentContextFactory;
+        private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+        private readonly RazorDocumentMappingService _documentMappingService;
+        private readonly ClientNotifierServiceBase _languageServer;
+        private readonly ILogger<RazorHoverEndpoint> _logger;
+
+        protected AbstractRazorDelegatingEndpoint(
+            DocumentContextFactory documentContextFactory,
+            LanguageServerFeatureOptions languageServerFeatureOptions,
+            RazorDocumentMappingService documentMappingService,
+            ClientNotifierServiceBase languageServer,
+            ILoggerFactory loggerFactory)
+        {
+            _documentContextFactory = documentContextFactory ?? throw new ArgumentNullException(nameof(documentContextFactory));
+            _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+            _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+            _logger = loggerFactory?.CreateLogger<RazorHoverEndpoint>() ?? throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        /// <summary>
+        /// Override to provide fallback behavior for when single server is not supported
+        /// </summary>
+        protected abstract Task<TResponse?> HandleWithoutSingleServerAsync(TRequest request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Override to provide behavior for razor contexts. Since the c#/html servers won't
+        /// have any information for this context it doesn't need to be delegated
+        /// </summary>
+        protected abstract Task<TResponse?> HandleInRazorAsync(TRequest request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken);
```

Comment:
Could probably simplify the API and require the consumer to retrieve the code document / source text (it'll be cached anyways)

```suggestion
        protected abstract Task<TResponse?> HandleInRazorAsync(TRequest request, DocumentContext documentContext, CancellationToken cancellationToken);
```

Reply:
wasn't sure about it being cached. In that case could probably just follow same signature as `Handle` and let everything else be retrieved as needed